src/java/com/puppetlabs/puppetserver/pool/JRubyPool.java

```diff
@@ -1,234 +1,371 @@
-    * to consumers of this class adding an instance to the queue and adding it to
-    * the set of registered elements is visible roughly atomically. `synchronize`
-    * guarantees that registration is atomic, so (ignoring borrows happening)
-    * `RegisteredElements` and `liveQueue` can only diverge by 1. "This is only
-    * "roughly" atomic, as the underlying queue uses its own lock, so it is
-    * possible for it to be modified on another thread while this method is being
-    * executed.
-    *
-    * @param e the element to register and put at the end of the queue.
-    *
-    * @throws InterruptedException
-    */
-    @Override
-    synchronized public void register(E e) throws InterruptedException {
-        registeredElements.add(e);
-        liveQueue.putLast(e);
-    }
+    // Lock which guards all accesses to the underlying queue and registered
+    // element set.  Constructed as "nonfair" for performance, like the lock
+    // that a `LinkedBlockingDeque` does.  Not clear that we need this
+    // to be a "fair" lock.
+    private final ReentrantLock lock = new ReentrantLock();
+
+    // Condition signaled when all elements that have been registered have been
+    // returned to the queue.  Awaited when a lock has been requested but
+    // one or more registered elements has been borrowed from the pool.
+    private final Condition allRegisteredInQueue = lock.newCondition();
+
+    // Condition signaled when an element has been added into the queue.
+    // Awaited when a request has been made to borrow an item but no elements
+    // currently exist in the queue.
+    private final Condition notEmpty = lock.newCondition();
+
+    // Condition signaled when the pool has been unlocked.  Awaited when a
+    // request has been made to borrow an item or lock the pool but the pool
+    // is currently locked.
+    private final Condition notWriteLocked = lock.newCondition();
+
+    // Holds a reference to all of the elements that have been registered.
+    // Newly registered elements are also added into the `liveQueue`.
+    // Elements only exist in the `liveQueue` when not currently
+    // borrowed whereas elements that have been registered (but not
+    // yet unregistered) will be accessible via `registeredElements`
+    // even while they are borrowed.
+    private final Set<E> registeredElements = new CopyOnWriteArraySet<>();
+
+    // Maximum size that the underlying queue can grow to.
+    private int maxSize;
 
+    // Thread which currently holds the write lock.  null indicates that
+    // there is no current write lock holder.  Using the current Thread
+    // object for tracking the lock owner is comparable to what the JDK's
+    // `ReentrantLock` class does:
+    //
+    // http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/util/concurrent/locks/ReentrantLock.java#l164
+    private Thread writeLockThread = null;
```

Comment:
I think `poolLockThread` is a pretty good name for this variable.

For the other variable ... hmm ... that one is harder ... I'd prefer something with the word "queue" instead of "deque" to match the convention established by the other variables in this class, like `liveQueue`.  So, some ideas ...
* `queueLock`
* `queueAndElementSetLock`
* `elementsLock`
* `internalLock`

And just to be contrarian, because why not, I don't hate just leaving it as `lock` since it is fundamental to the implementation, and that's what the JDK classes do (I think), and `poolLockThread` seems to distinguish it from this lock.

If I had to choose it would probably be `lock` or `queueLock`.